docs/design.md

```diff
@@ -0,0 +1,785 @@
+metadata:
+  schema: metadata/Document/v1
+  name: global-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: global
+data:
+  a:
+    x: 1
+    y: 2
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: region-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: region
+    parentSelector:
+      key1: value1
+    actions:
+      - method: replace
+        path: .a
+data:
+  a:
+    z: 3
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: site-1234
+  layeringDefinition:
+    layer: site
+    parentSelector:
+      key1: value1
+    actions:
+      - method: merge
+        path: .
+data:
+  b: 4
+...
+```
+
+When rendering, the parent chosen for `site-1234` will be `region-1234`,
+since it is the highest precedence document that matches the label selector
+defined by `parentSelector`, and the parent chosen for `region-1234` will be
+`global-1234` for the same reason. The rendered result for `site-1234` would
+be:
+
+```yaml
+---
+schema: example/Kind/v1
+metadata:
+  name: site-1234
+data:
+  a:
+    z: 3
+  b: 4
+...
+```
+
+If `region-1234` were later removed, then the parent chosen for `site-1234`
+would become `global-1234`, and the rendered result would become:
+
+```yaml
+---
+schema: example/Kind/v1
+metadata:
+  name: site-1234
+data:
+  a:
+    x: 1
+    y: 2
+  b: 4
+...
+```
+
+<!-- TODO: Add figures for this example, with region present, have site point
+with dotted line at global and indicate in caption (or something) that it's
+selected for but ignored, because there's a higher-precedence layer to select
+-->
+
+### Substitution
+
+Substitution is primarily designed as a mechanism for inserting secrets into
+configuration documents, but works for unencrypted source documents as well.
+Substitution is applied at each layer after all merge actions occur.
+
+Concrete (non-abstract) documents can be used as a source of substitution
+into other documents. This substitution is layer-independent, so given the 3
+layer example above, which includes `global`, `region` and `site` layers, a
+document in the `region` layer could insert data from a document in the
+`site` layer.
+
+Here is a sample set of documents demonstrating subistution:
+
+```yaml
+---
+schema: deckhand/Certificate/v1
+metadata:
+  name: example-cert
+  storagePolicy: cleartext
+  layeringDefinition:
+    layer: site
+data: |
+  CERTIFICATE DATA
+---
+schema: deckhand/CertificateKey/v1
+metadata:
+  name: example-key
+  storagePolicy: encrypted
+  layeringDefinition:
+    layer: site
+data: |
+  KEY DATA
+---
+schema: deckhand/Passphrase/v1
+metadata:
+  name: example-password
+  storagePolicy: encrypted
+  layeringDefinition:
+    layer: site
+data: my-secret-password
+---
+schema: armada/Chart/v1
+metadata:
+  name: example-chart-01
+  storagePolicy: cleartext
+  layeringDefinition:
+    layer: region
+  substitutions:
```

Comment:
Substitutions are applied in order and the layer of the source is ignored (except that it must be first fully rendered before it's used as a substitution source).

This means that there can in fact be conflict.  I think that since it's completely deterministic we might not want to forbid it right out the gate, thought I could quickly be talked into a deckhand-provided validation that checks for such scenarios.